### ISSUE TITLE
recalc total VAT for rate after correcting VAT for one item

### ISSUE THREAD

**Issue by jstaerk** [kyr0/easy-erechnung#16]
when recognizing the [mustang sample invoice](https://www.mustangproject.org/files/MustangGnuaccountingBeispielRE-20190610_507blanko.pdf) the item with 7% VAT is recognized as having 19%. This can be manually adjusted in the correction step, but then also the tax rates need to be recalculated, i.e. the according amount has to be listed under the according rax rate. Mustang-wise you can probably use transactioncalculator for that.

**Comment by kyr0** [comment]
On a side-note: Regarding the re-calculation, I wonder if the goal of the app should be to "interpret/validate/correct" or to "believe/transform"? My guts feeling is that it would be ideal for this app to focus on 100% correct OCR, believing that the data read from the images (PDFs are turned into images before OCRing them) is correct, interpret only the layout/data and semantic context for mapping to the ZUGFeRD/Factur-X standard? I fear that the moment we start "re-calculating" in the app, the app is taking "decisions" it might be "too dumb to answer" - especially taking the possibility of OCR errors into consideration. People might not "see the error" and believe in the numbers without a second thought - ending up filing invoices to customers and tax reports they never intended this way. 

I'm really wondering what the best way to solve this issue would look like? Technically, I'm working on a new implementation of the OCR process that involves, for each section, a `confidence` value between `0..1`. Would the value be < 0.7, I would mark the input field where we render the "uncertain" value into with a bold red border.

Should we decide for recalculating values, we could use a bold yellow border to inform the user. In a way like: "Hey look, these values require your attention!"

What do you think?

**Comment by jstaerk** [comment]
Hi

> I'm really wondering what the best way to solve this issue would look like? 

Fundamentally, if the invoice it not correct, it should have been rejected in the first place, but since recognition is not 100% the only errors we need to be able to correct are recognition errors. But yes, automatic recalculation should be kept to a minimum, yes, indicating uncertainty sounds like a good idea and yes, yellow sounds logical.

Can the redundancy of an invoice help with the recognition? E.g. if you have 100€ in a 7% block in the VAT totals could one finetune the AI to assume a higher likelyhood for the only 100€-line to be of 7% VAT?

**Comment by kyr0** [comment]
> Fundamentally, if the invoice it not correct, it should have been rejected in the first place, but since recognition is not 100% the only errors we need to be able to correct are recognition errors. But yes, automatic recalculation should be kept to a minimum, yes, indicating uncertainty sounds like a good idea and yes, yellow sounds logical.

Great, then we are on the same page.

> Can the redundancy of an invoice help with the recognition? E.g. if you have 100€ in a 7% block in the VAT totals could one finetune the AI to assume a higher likelyhood for the only 100€-line to be of 7% VAT?

Partly. As the AI-based OCR process is usually a two-step process, as in 1) OCR model transforming image to text description (Markdown usually) followed by 2) LLM transforming Markdown to JSON, the LLM can be tuned via the prompt to pay attention to such redundancy. 

However, for finetuning, without a huge set of invoices, speaking of a few thousand correctly labeled ones, a fine-tune tends to overfit the model (=> catastrophic quality loss). If we want to have a great OCR model for European invoices, we need tons of correctly labled invoices - there is no way around it that I know of.

The few labeled invoices that I have (also thanks to you, @jstaerk) and a few others are basically invoices that I can use "privately" for evaluation purposes only. This helps, but it doesn't help to solve the root problem for now, unfortunately.
